lldb/source/Commands/CommandOptionsProcessLaunch.cpp

```diff
@@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
   case 's': // Stop at program entry point
     launch_info.GetFlags().Set(eLaunchFlagStopAtEntry);
     break;
-
+  case 'm': // Stop at main function
+  {
+    TargetSP target_sp =
+        execution_context ? execution_context->GetTargetSP() : TargetSP();
+    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
+        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
```

Comment:
IIUC, we're not trying to break at the actual entry point (that's what -stop-at-entry-point does) but the main executable's Language entry point.  I'm not sure that is something the platform can know.   For all the C-based languages it was "main", but for some other languages it might be something else.  For Ada apparently it can be anything and you designate it "main" some other way.  In other languages it could just have a different name than "main"...

If that's true, shouldn't we be asking all the Languages we support what their "main" function is, not the Platforms.  I'm not sure they can know.  And for situations like the Ada one the language would have to be able to figure out what the main function was by some Ada specific means.

Reply:
In the simplest case, we gather a set of names one or more from each Language, and then make one multiple name breakpoint for all of them.

Reply:
Yeah, there is no way to know what language the main entrypoint was written in, so no hope of figuring out the language. I would be fine with the multiple names approach from all languages. If you want that feature, can you give Jose some clues what he should do to make that happen?

Reply:
We have a list of "Supported Languages" (Language::GetSupportedLangauges) which are the ones we actually have plugins for as opposed to the ones whose names we know.  That's the list you should be querying. 

We should add a new API to the Language plugin: Language::GetUserEntrypointName() and then this code should iterate over all the supported languages building up a unique list of entry point names using this API.  At present there's going to be one: "main" repeated many times but I think Walter promised us another one soon...  

If you find more than one name, there's a Target::CreateBreakpoint overload that takes a `const char *func_names[]` - that you can use to make one portmanteau breakpoint for all the entry point names.  

It's also a good idea to add a shared library filter instead of the first `nullptr` in the code above, and specify the main executable.  You know the entry point has to be in the main executable, and so it would be inefficient to search the whole world for this breakpoint.  It could also end up not being what you want, because some other shared library might call a function named "main" as part of it's shared library loading code (which happens before the language's entry point) so you would in fact stop too early.

You'll need to pass eLanguageTypeUnknown for the language.

Reply:
It's probably also fine to implement a default version in the Language class that returns `main`.  No need to make everyone override this method when it's `main` for most languages.

Reply:
> We should add a new API to the Language plugin: Language::GetUserEntrypointName() and then this code should iterate over all the supported languages building up a unique list of entry point names using this API. At present there's going to be one: "main" repeated many times but I think Walter promised us another one soon...

About this, Walter had proposed the following: "I'd suggest adding a virtual method GetMainSymbol to the Language.h interface, which could be implemented by each language plugin.". I guess it's a good approach. What are your thoughts?

Reply:
> It's also a good idea to add a shared library filter instead of the first `nullptr` in the code above, and specify the main executable. You know the entry point has to be in the main executable, and so it would be inefficient to search the whole world for this breakpoint. It could also end up not being what you want, because some other shared library might call a function named "main" as part of it's shared library loading code (which happens before the language's entry point) so you would in fact stop too early.

I see. Something like this?

```cpp
ModuleSP main_module_sp = target_sp->GetExecutableModule();
FileSpecList shared_lib_filter;
shared_lib_filter.Append(main_module_sp->GetFileSpec());
BreakpointSP bp_sp = target_sp->CreateBreakpoint(
    &shared_lib_filter, nullptr, "main", eFunctionNameTypeFull, eLanguageTypeUnknown,
    0, eLazyBoolNo, false, false);
```